---
title: "Analyzing single-cell RNA-seq count data with Celda"
author: "Josh D. Campbell, Masanao Yajima, Sean Corbett, Zichun Liu, Shiyi Yang, Tianwen Huan, Anahita Bahri, Zhe Wang, Yusuke Koga"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
## 1 Introduction
  
  "celda" stands for "**CE**llular **L**atent **D**irichlet **A**llocation", which is a suite of Bayesian hierarchical models and supporting functions to perform gene and cell clustering for count data generated by single cell RNA-seq platforms. This algorithm is an extension of the Latent Dirichlet Allocation (LDA) topic modeling framework that has been popular in text mining applications. Celda has advantages over other clustering frameworks:
  
1) Celda can simultaneously cluster cells into subpopulations and genes into transcriptional states, which are subgroups of genes that are likely to be expressed in similar proportions over different types of cells
2) Celda uses count-based Dirichlet-multinomial distributions so no additional normalization is required for 3' DGE single cell RNA-seq
3) These types of models have shown good performance with sparse data.

In this vignette we will demonstrate how to perform cell and gene clustering with simulated and real single cell RNA-seq data using the Bayesian hierarchical models within celda. 


## Installation Instructions

Currently `celda` is hosted on a Github repository `compbiomed/celda`.

```{r,eval=FALSE}
library(devtools)
install_github("compbiomed/celda")
```

This should automatically install `celda` on your local machine as long as you have proper internet connection.  If you encounter a problem during the installation you may want to open up a browser and acces our github repository https://github.com/compbiomed/celda.

## Loading the library

Once installed the package can be loaded using the `library` command.

```{r,eval=FALSE}
library(celda)
```

## Finding the helpfiles
Complete list of help files are accessible using the help command with a `package` option.

```{r,eval=FALSE}
help(package=celda)
```


# 2. Single cell RNA-seq dataset analysis with celda

We will learn how to use the package `celda` to cluster genes and cells from count data generated in single cell RNA-seq experiments. `celda` will take a matrix of counts where each row is a gene and each cell is a column and each entry in the matrix is the number of counts for each gene in each cell. For this tutorial, we will start with a simple matrix of counts that were simulated according to the generative process behind the `celda_CG` model and then we will apply `celda` to a single-cell RNA-seq dataset of PBMCs.  

```{r,eval=T, warning = FALSE, echo = FALSE}
library(devtools)
library(celda)
library(Matrix)
library(gtools)
library(ggplot2)
library(tsne)
library(reshape2)
```

We use the built-in data generating function `simulateCells.celda_CG` to simulate a single-cell RNA-Seq dataset. `simulateCells.celda_CG` returns a list containing a count matrix where each row is a gene and each column is a cell. The K parameter determines the number of cell clusters while the L parameter determines the number of gene clusters within the dataset.

```{r}
sim_counts = simulateCells(K = 3, L = 10,model = "celda_CG")

```

The `counts` variable from `simulateCells.celda_CG` contains the counts matrix.
The dimensions of counts matrix:

```{r}
dim(sim_counts$counts)
```

The `z` variable from `simulateCells.celda_CG` contains the cluster for each cell.
Here is the number of cells in each subpopulation:

```{r}
table(sim_counts$z)
```

The `y` variable from `simulateCells.celda_CG` contains the transcriptional state assignment for each gene. 
Here is the number of genes in each transcriptional state:

```{r}
table(sim_counts$y)
```

The sample.label is used to denote the sample from which each cell was derived. In this simulated dataset, we have 10 samples:

```{r}
table(sim_counts$sample.label)
```

Each cell is assumed to come from a sample. Here is the number of cells in each subpopulation within each sample:

```{r}
table(sim_counts$z, sim_counts$sample.label)
```



## 2.1.1 How to Run Celda

There are currently three models within this package: `celda_C` will cluster cells, `celda_G` will cluster genes, and `celda_CG` will simultaneously cluster cells and genes. Here is the command for `celda_CG`:

```{r, fig.show='hold', warning = FALSE, message = FALSE}

celdaCG.res = celda_CG(counts = sim_counts$counts, K = 3, L = 10, max.iter = 10, sample.label = sim_counts$sample.label)
```

Here is a comparison between the true cluster labels and the estimated cluster labels.

```{r}
z = celdaCG.res$z
y = celdaCG.res$y

table(z, sim_counts$z)
table(y, sim_counts$y)

```

We can display the clustering results with a heatmap of the normalized counts:
 c
```{r, fig.width = 7, fig.height = 7, warning = FALSE, message = FALSE}
norm.counts <- normalizeCounts(sim_counts$counts)
render_celda_heatmap(counts = norm.counts, z=z, y=y, normalize = NULL, color_scheme = "divergent",cluster_gene = TRUE, cluster_cell = TRUE)
```

## 2.1.2 Matrix factorization

`celda` can also perform matrix factorization to summarize the contribution of each transcriptional state to each cellular subpopulation. 

```{r}
factorized <- factorizeMatrix.celda_CG(counts = sim_counts$counts, celda.obj = celdaCG.res)

pop.states = t(factorized$proportions$population.states)
head(pop.states)
```

`celda` contains several plotting tools for data dimensionality reduction tool outputs. The PCA result for the factorized matrix is plotted by the `plor_dr_cluster` and `plot_dr_state` function based off of the celda clusters and state probabilities for each cell:

```{r, fig.width = 7, fig.height = 7}
data.pca <- prcomp(t(scale(t(factorized$proportions$cell.states))),scale = F, center = F)

plot_dr_cluster(dim1 = data.pca$rotation[,1], dim2 = data.pca$rotation[,2], cluster = as.factor(celdaCG.res$z))

plot_dr_state(dim1 = data.pca$rotation[,1], dim2 = data.pca$rotation[,2], matrix = factorized$proportions$cell.states, rescale = TRUE)
```

The transcriptional and cell states can also be visualized with a heatmap:

```{r, fig.width = 7, fig.height = 7}
render_celda_heatmap(pop.states, color_scheme = "sequential", show_cellnames=T, show_genenames=T, breaks = NA, z= 1:3, y = 1:10)
```

Sometimes it may be helpful to visualize the relative probability of each transcriptional state in each cellular subpopulation. We can do this by z-score normalizing the probability matrix:

```{r, fig.width = 7, fig.height = 7}
rel.states = sweep(pop.states, 1, rowSums(pop.states), "/")
render_celda_heatmap(rel.states, color_scheme = "sequential", show_cellnames=T, show_genenames=T, breaks = NA, z= 1:3, y = 1:10)
```

The relative probabilities will tend to reflect the patterns observed in the heatmap as the genes in the heatmaps are Z-score transformed and thus on a relative scale as well.


## 2.1.3 Identifying the optimal number of states

The optimal number of K and L will likely not be known a priori. Therefore multiple choices of K and L may need to tested and compared. Additionally, `celda` is sensitive to initial start conditions, so it is good practice to run multiple chains for each combination of K and L to increase the chances of finding the most optimal solution. We have built a wrapper function that will run multiple combinations of K and L with multiple chains for each combination in parallel. 

```{r}
celda.res.list <- celda(counts = sim_counts$counts,K = 2:4, L = 9:11, nchains = 3, max.iter = 10, model = "celda_CG")
```

Here, `celda` will fit the model for every combination of K between 2 and 4 and L between 9 and 11. Three chains will be run for each combination. The parameter "ncores" can be set to something more than 1 to run multiple chains in parallel.

We can use the function "getModel" to select a particular model from the list of models once we have selected the optimal K and L.

```{r}
model = getModel(celda.res.list, K = 3, L = 10, best = "loglik")
```

The chain with the best log likelihood will be returned from all models that were run with K = 3 and L = 10.

# 3. PBMC Analysis with celda

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.align = "center")
library(celda)
library(Matrix)
library(data.table)
library(pheatmap)
```

We will use the "cmatp_data" dataset, which is a dataset of 3000 Peripheral Blood Mononuclear Cells (PBMC), available from 10X Genomics. The rows are organized by gene names, while the columns are organized by barcodes. For this tutorial, the dataset has been slightly modified so the rownames are comprised of the Ensembl gene ID as well as the gene name. The raw dataset can be found at https://support.10xgenomics.com/single-cell/software/pipelines/latest/rkit.

We will only include genes with at least 5 counts in a least 5 cells for this analysis to reduce computational time. However, `celda` is capable of analyzing genes with fewer counts. 

```{r}
load(file="../data/cmatp_data.rda")
dim(cmatp_data)
head(rownames(cmatp_data))
head(colnames(cmatp_data))
```

```{r}
cmatp_select <- cmatp_data[rowSums(cmatp_data>4) > 4,]
```

### 3.1.1 Analysis of 10X Genomics PBMC data


```{r, message = FALSE, verbose = FALSE}
cmatp.res = celda(cmatp_select, K = 15, L = 20, model = "celda_CG", nchains = 3, cores = 1)
model.res.list<-getModel(cmatp.res, K = 15, L = 20)
```

After matrix factorization, the top genes can be selected using the `topRank` function on the "gene.states" matrix:  

```{r}
factorize.matrix = factorizeMatrix.celda_CG(counts = cmatp_select, celda.obj = model.res.list)
top.genes <- topRank(factorize.matrix$proportions$gene.states, n = 25)
```

```{r}
top.genes$names$L16
```

```{r}
top.genes$names$L20
```

We can make a heatmap of these top genes as follows: 

```{r, fig.width = 7, fig.height = 7}
top.genes.ix <- unique(unlist(top.genes$index))
norm.cmatp.counts <- normalizeCounts(cmatp_select)
render_celda_heatmap(norm.cmatp.counts[top.genes.ix,], z = model.res.list$z, y = model.res.list$y[top.genes.ix], normalize = NULL, color_scheme = "divergent")
```

We can plot tSNE results via `plot_dr_cluster`, `plot_dr_state` and `plot_dr_gene`. These will determine how each cell is labeled via celda, visualize each cell's expression of a specific transcriptional state and visualize each cell's expression of a set of genes, respectively. 

```{r, message = FALSE}
norm_cmatp <- normalizeCounts(factorize.matrix$counts$cell.states)
set.seed(123)
cmatp_tsne <- tsne(t(norm_cmatp), k = 2)

```

```{r, fig.width = 7, fig.height = 7}
plot_dr_cluster(dim1 = cmatp_tsne[,1], dim2 = cmatp_tsne[,2], cluster = as.factor(model.res.list$z))

plot_dr_state(dim1 = cmatp_tsne[,1], dim2 = cmatp_tsne[,2], matrix = factorize.matrix$proportions$cell.states, rescale = TRUE)

marker.genes <- c("ENSG00000179639_FCER1A","ENSG00000105374_NKG7","ENSG00000153563_CD8A" ,"ENSG00000168685_IL7R", "ENSG00000198851_CD3E", "ENSG00000156738_MS4A1", "ENSG00000203747_FCGR3A","ENSG00000163736_PPBP","ENSG00000090382_LYZ")
gene.counts <- cmatp_select[marker.genes,]
plot_dr_gene(dim1 = cmatp_tsne[,1],dim2 = cmatp_tsne[,2], matrix = gene.counts, rescale = TRUE)
```

With the use of the plots, it is also possible to determine which genes express sepcific marker genes. For instance, the same cell cluster that highly expresses the transcriptional state L20 highly expresses PPBP, a known marker gene for megakaryocytes. This is consistent with what we see in the above, in which transcriptional state L20 highly expresses PPBP.