<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of single-cell genomic data with celda • celda</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of single-cell genomic data with celda">
<meta property="og:description" content="celda">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">celda</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.10.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/celda">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/articles/celda_pbmc3k.html">Celda - Analysis of PBMC3K</a>
    </li>
    <li>
      <a href="../articles/articles/decontX_pbmc4k.html">DecontX - Decontamination of PBMC4K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/campbio/celda">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/campbio/celda/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="celda_files/header-attrs-2.7/header-attrs.js"></script><script src="celda_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of single-cell genomic data with celda</h1>
                        <h4 class="author">Joshua Campbell</h4>
            <address class="author_afil">
      Boston University School of Medicine<br><a class="author_email" href="mailto:#"></a><a href="mailto:camp@bu.edu" class="email">camp@bu.edu</a>
      </address>
                              <h4 class="author">Zhe Wang</h4>
            <address class="author_afil">
      Boston University School of Medicine<br><h4 class="author">Shiyi Yang</h4>
            <address class="author_afil">
      Boston University School of Medicine<br><h4 class="author">Sean Corbett</h4>
            <address class="author_afil">
      Boston University School of Medicine<br><h4 class="author">Yusuke Koga</h4>
            <address class="author_afil">
      Boston University School of Medicine<br><h4 class="date">2021-12-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/campbio/celda/blob/master/vignettes/celda.Rmd"><code>vignettes/celda.Rmd</code></a></small>
      <div class="hidden name"><code>celda.Rmd</code></div>

    </address>
</address>
</address>
</address>
</div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><strong>CE</strong>llular <strong>L</strong>atent <strong>D</strong>irichlet <strong>A</strong>llocation (celda) is a collection of Bayesian hierarchical models to perform feature and cell bi-clustering for count data generated by single-cell platforms. This algorithm is an extension of the Latent Dirichlet Allocation (LDA) topic modeling framework that has been popular in text mining applications and has shown good performance with sparse data. celda simultaneously clusters features (i.e. gene expression) into modules based on co-expression patterns across cells and cells into subpopulations based on the probabilities of the feature modules within each cell.</p>
<p>Starting from Bioconductor release 3.12 (<code>celda</code> version 1.6.0), <code>celda</code> makes use of <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment">SingleCellExperiment</a></em> (SCE) objects for storing data and results. In this vignette we will demonstrate how to use celda to perform cell and feature clustering with a simple, small simulated dataset. This vignette does not include upstream importing of data, quality control, or filtering. To see a more complete analysis of larger real-world datasets, visit <a href="https://www.camplab.net/celda/">camplab.net/celda</a> for additional vignettes.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>celda can be installed from Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"celda"</span><span class="op">)</span></code></pre></div>
<p>To load the package, type the following:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">celda</span><span class="op">)</span></code></pre></div>
<p>A complete list of help files are accessible using the help command with the <code>package</code> option.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/help.html">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="va">celda</span><span class="op">)</span></code></pre></div>
<p>To see the latest updates and releases or to post a bug, see our GitHub page at <a href="https://github.com/campbio/celda" class="uri">https://github.com/campbio/celda</a>. To ask questions about running celda, post a thread on Bioconductor support site at <a href="https://support.bioconductor.org/" class="uri">https://support.bioconductor.org/</a>.</p>
<p><br></p>
</div>
<div id="generation-of-a-simulated-single-cell-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#generation-of-a-simulated-single-cell-dataset" class="anchor"></a>Generation of a simulated single cell dataset</h1>
<p>celda will take a matrix of counts where each row is a feature, each column is a cell, and each entry in the matrix is the number of counts of each feature in each cell. To illustrate the utility of celda, we will apply it to a simulated dataset.</p>
<p>In the function <code>simulateCells</code>, the <strong>K</strong> parameter designates the number of cell clusters, the <strong>L</strong> parameter determines the number of feature modules, the <strong>S</strong> parameter determines the number of samples in the simulated dataset, the <strong>G</strong> parameter determines the number of features to be simulated, and <strong>CRange</strong> specifies the lower and upper bounds of the number of cells to be generated in each sample.</p>
<p>To simulate a dataset of 5 samples with 5 cell populations, 10 feature modules, 200 features, and between 30 to 50 cells per sample using <code>celda_CG</code> model:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simsce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulateCells.html">simulateCells</a></span><span class="op">(</span><span class="st">"celda_CG"</span>,
    S <span class="op">=</span> <span class="fl">5</span>, K <span class="op">=</span> <span class="fl">5</span>, L <span class="op">=</span> <span class="fl">10</span>, G <span class="op">=</span> <span class="fl">200</span>, CRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The <code>counts</code> assay slot in <code>simsce</code> contains the counts matrix. The dimensions of counts matrix:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 200 207</code></pre>
<p>Columns <code>celda_sample_label</code> and <code>celda_cell_cluster</code> in <code>colData(simsce)</code> contain sample labels and celda cell population cluster labels. Here are the numbers of cells in each subpopulation and in each sample:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">$</span><span class="va">celda_cell_cluster</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  1  2  3  4  5 
## 42 44 40 47 34</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">$</span><span class="va">celda_sample_label</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 
##       43       48       45       40       31</code></pre>
<p>Column <code>celda_feature_module</code> in <code>rowData(simsce)</code> contains feature module labels. Here is the number of features in each feature module:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">$</span><span class="va">celda_feature_module</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  1  2  3  4  5  6  7  8  9 10 
## 23 39 17 15 21 22 19 12  4 28</code></pre>
</div>
<div id="feature-selection" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-selection" class="anchor"></a>Feature selection</h1>
<p>A simple heuristic feature selection is performed to reduce the size of features used for clustering. To speed up the process, only features with at least 3 counts in at least 3 cells are included in downstream clustering for this data. A subset <code>SingleCellExperiment</code> object with filtered features is stored in <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp(simsce, "featureSubset")</a></code> slot by default.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simsce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectFeatures.html">selectFeatures</a></span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span></code></pre></div>
<p>If the number of features is still too large, then a smaller subset of features can be obtained by selecting the top number of most variable genes. For an example code, see the PBMC3K tutorial in the online celda <a href="https://www.camplab.net/celda">documentation</a>.</p>
</div>
<div id="performing-bi-clustering-with-celda" class="section level1">
<h1 class="hasAnchor">
<a href="#performing-bi-clustering-with-celda" class="anchor"></a>Performing bi-clustering with celda</h1>
<p>There are currently three models within <em>celda</em> package: <code>celda_C</code> will cluster cells, <code>celda_G</code> will cluster features, and <code>celda_CG</code> will simultaneously cluster cells and features. Within the functions the <code>K</code> parameter will be the number of cell populations to be estimated, while the <code>L</code> parameter will be the number of feature modules to be estimated in the output model.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celda_CG.html">celda_CG</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">simsce</span>, K <span class="op">=</span> <span class="fl">5</span>, L <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, nchains <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Here is a comparison between the true cluster labels and the estimated cluster labels.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/celdaClusters.html">celdaClusters</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fu"><a href="../reference/celdaClusters.html">celdaClusters</a></span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##    
##      1  2  3  4  5
##   1  0 44  0  0  0
##   2 42  0  0  0  0
##   3  0  0 40  0  0
##   4  0  0  0 47  0
##   5  0  0  0  0 34</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/celdaModules.html">celdaModules</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fu"><a href="../reference/celdaModules.html">celdaModules</a></span><span class="op">(</span><span class="va">simsce</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##     
##       1  2  3  4  5  6  7  8  9 10
##   1   0 32  0  0  0  0  0  0  0  0
##   2  19  0  0  0  0  0  0  0  0  0
##   3   0  0 15  0  0  0  0  0  0  0
##   4   0  0  0 13  0  0  0  0  0  0
##   5   0  0  0  0 21  0  0  0  0  0
##   6   0  0  0  0  0 19  0  0  0  0
##   7   0  0  0  0  0  0  0 12  0  0
##   8   0  0  0  0  0  0 17  0  0  0
##   9   0  0  0  0  0  0  0  0  3  0
##   10  0  0  0  0  0  0  0  0  0 20</code></pre>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
<div id="plotting-cell-populations-on-2d-embeddings" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-cell-populations-on-2d-embeddings" class="anchor"></a>Plotting cell populations on 2D-embeddings</h2>
<p>celda contains its own wrapper function for tSNE and UMAP called <code>celdaTsne</code> and <code>celdaUmap</code>, respectively. Both of these functions can be used to embed cells into 2-dimensions. The output can be used in the downstream plotting functions <code>plotDimReduceCluster</code>, <code>plotDimReduceModule</code>, and <code>plotDimReduceFeature</code> to show cell population clusters, module probabilities, and expression of individual features, respectively.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celdaUmap.html">celdaUmap</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDimReduceCluster.html">plotDimReduceCluster</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_UMAP"</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/plot_umap-1.png" width="672"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDimReduceModule.html">plotDimReduceModule</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_UMAP"</span>, rescale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/plot_umap-2.png" width="672"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDimReduceFeature.html">plotDimReduceFeature</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_UMAP"</span>,
    normalize <span class="op">=</span> <span class="cn">TRUE</span>, features <span class="op">=</span> <span class="st">"Gene_1"</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/plot_umap-3.png" width="672"></p>
</div>
<div id="creating-an-expression-heatmap" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-expression-heatmap" class="anchor"></a>Creating an expression heatmap</h2>
<p>The clustering results can be viewed with a heatmap of the normalized counts using the function <code>celdaHeatmap</code>. The top <code>nfeatures</code> in each module will be selected according to the factorized module probability matrix.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/celdaHeatmap.html">celdaHeatmap</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce</span>, nfeatures <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/celda_heatmap-1.png" width="672"></p>
</div>
<div id="displaying-relationships-between-modules-and-cell-populations" class="section level2">
<h2 class="hasAnchor">
<a href="#displaying-relationships-between-modules-and-cell-populations" class="anchor"></a>Displaying relationships between modules and cell populations</h2>
<p>The relationships between feature modules and cell populations can be visualized with <code>celdaProbabilityMap</code>. The absolute probabilities of each feature module in each cellular subpopulation is shown on the left. The normalized and z-scored expression of each module in each cell population is shown on the right.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/celdaProbabilityMap.html">celdaProbabilityMap</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/propmap-1.png" width="672"></p>
</div>
<div id="examining-co-expression-with-module-heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#examining-co-expression-with-module-heatmaps" class="anchor"></a>Examining co-expression with module heatmaps</h2>
<p><code>moduleHeatmap</code> creates a heatmap using only the features from a specific feature module. Cells are ordered from those with the lowest probability of the module to the highest. If more than one module is used, then cells will be ordered by the probabilities of the first module.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/moduleHeatmap.html">moduleHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, featureModule <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, topCells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/module_heatmap-1.png" width="672"></p>
</div>
</div>
<div id="identifying-reasonable-numbers-of-feature-modules-and-cell-subpopulations" class="section level1">
<h1 class="hasAnchor">
<a href="#identifying-reasonable-numbers-of-feature-modules-and-cell-subpopulations" class="anchor"></a>Identifying reasonable numbers of feature modules and cell subpopulations</h1>
<p>In the previous example, the best <code>K</code> (the number of cell clusters) and <code>L</code> (the number of feature modules) was already known. However, the optimal <code>K</code> and <code>L</code> for each new dataset will likely not be known beforehand and multiple choices of <code>K</code> and <code>L</code> may need to be tried and compared. celda offers two sets of functions to determine the optimum <code>K</code> and <code>L</code>, <code>recursiveSplitModule</code>/<code>recursiveSplitCell</code>, and <code>celdaGridSearch</code>.</p>
<div id="using-recursive-splitting" class="section level2">
<h2 class="hasAnchor">
<a href="#using-recursive-splitting" class="anchor"></a>Using recursive splitting</h2>
<p>Functions <code>recursiveSplitModule</code> and <code>recursiveSplitCell</code> offer a fast method to generate a celda model with optimum <code>K</code> and <code>L</code>. First, <code>recursiveSplitModule</code> is used to determine the optimal <code>L</code>. <code>recursiveSplitModule</code> first splits features into however many modules are specified in <code>initialL</code>. The module labels are then recursively split in a way that would generate the highest log-likelihood, all the way up to <code>maxL</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">moduleSplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recursiveSplitModule.html">recursiveSplitModule</a></span><span class="op">(</span><span class="va">simsce</span>, initialL <span class="op">=</span> <span class="fl">2</span>, maxL <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p>Perplexity is a statistical measure of how well a probability model can predict new data. Lower perplexity indicates a better model. The perplexity of each model can be visualized with <code>plotGridSearchPerplexity</code>. In general, visual inspection of the plot can be used to select the optimal number of modules (<code>L</code>) or cell populations (<code>K</code>) by identifying the “elbow” - where the rate of decrease in the perplexity starts to drop off.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGridSearchPerplexity.html">plotGridSearchPerplexity</a></span><span class="op">(</span><span class="va">moduleSplit</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>In this example, the perplexity for <code>L</code> stops decreasing at L = 10, thus L = 10 would be a good choice. Sometimes the perplexity alone does not show a clear elbow or “leveling off”. However, the rate of perplexity change (RPC) can be more informative to determine when adding new modules does not add much additional information <a href="https://doi.org/10.1186/1471-2105-16-S13-S8" target="_blank">Zhao et al., 2015</a>). An RPC closer to zero indicates that the addition of new modules or cell clusters is not substantially decreasing the perplexity. The RPC of models can be visualized using function <code>plotRPC</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRPC.html">plotRPC</a></span><span class="op">(</span><span class="va">moduleSplit</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/module_split_rpc-1.png" width="700"></p>
<p>Once you have identified the optimal <code>L</code> (in this case, L is selected to be 10), the module labels are used for initialization in <code>recursiveSplitCell</code>. Similarly to <code>recursiveSplitModule</code>, cells are initially split into a small number of subpopulations, and the subpopulations are recursively split up.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">moduleSplitSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetCeldaList.html">subsetCeldaList</a></span><span class="op">(</span><span class="va">moduleSplit</span>, params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>

<span class="va">cellSplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recursiveSplitCell.html">recursiveSplitCell</a></span><span class="op">(</span><span class="va">moduleSplitSelect</span>,
    initialK <span class="op">=</span> <span class="fl">3</span>,
    maxK <span class="op">=</span> <span class="fl">12</span>,
    yInit <span class="op">=</span> <span class="fu"><a href="../reference/celdaModules.html">celdaModules</a></span><span class="op">(</span><span class="va">moduleSplitSelect</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGridSearchPerplexity.html">plotGridSearchPerplexity</a></span><span class="op">(</span><span class="va">cellSplit</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/rpc_cell-1.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRPC.html">plotRPC</a></span><span class="op">(</span><span class="va">cellSplit</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/rpc_cell-2.png" width="700"></p>
<p>In this plot, the perplexity for K stops decreasing at K = 5, with a final K/L combination of K = 5, L = 10. Generally, this method can be used to pick a reasonable <code>L</code> and a potential range of <code>K</code>. However, manual review of specific selections of <code>K</code> is often required to ensure results are biologically coherent.</p>
<p>Once users have chosen the K/L parameters for further analysis, the <code>subsetCeldaList</code> function can be used to subset the celda list <em>SCE</em> object to a single model <em>SCE</em> object.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetCeldaList.html">subsetCeldaList</a></span><span class="op">(</span><span class="va">cellSplit</span>, params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fl">5</span>, L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="using-a-grid-search" class="section level2">
<h2 class="hasAnchor">
<a href="#using-a-grid-search" class="anchor"></a>Using a grid search</h2>
<p>Alternativley to recursive splitting, celda is able to run multiple combinations of K and L with multiple chains in parallel via the <code>celdaGridSearch</code> function.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celdaGridSearch.html">celdaGridSearch</a></span><span class="op">(</span><span class="va">simsce</span>,
    paramsTest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span>, L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span>,
    cores <span class="op">=</span> <span class="fl">1</span>,
    model <span class="op">=</span> <span class="st">"celda_CG"</span>,
    nchains <span class="op">=</span> <span class="fl">2</span>,
    maxIter <span class="op">=</span> <span class="fl">100</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span>,
    bestOnly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Setting <code>verbose</code> to <code>TRUE</code> will print the output of each model to a text file. These results can be visualized with <code>plotGridSearchPerplexity</code>. The major goal is to pick the lowest <code>K</code> and <code>L</code> combination with relatively good perplexity. In general, visual inspection of the plot can be used to select the number of modules (<code>L</code>) or cell populations (<code>K</code>) where the rate of decrease in the perplexity starts to drop off. <code>bestOnly = TRUE</code> indicates that only the chain with the best log likelihood will be returned for each K/L combination.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotGridSearchPerplexity.html">plotGridSearchPerplexity</a></span><span class="op">(</span><span class="va">cgs</span><span class="op">)</span></code></pre></div>
<p><img src="celda_files/figure-html/plot_grid_search-1.png" width="768"></p>
<p>In this example, the perplexity for <code>L</code> stops decreasing at L = 10 for the majority of <code>K</code> values. For the line corresponding to L = 10, the perplexity stops decreasing at K = 5. Thus L = 10 and K = 5 would be a good choice. Again, manual review of specific selections of K is often be required to ensure results are biologically coherent.</p>
<p>Once users have chosen the K/L parameters for further analysis, the <code>subsetCeldaList</code> function can be used to subset the celda list <em>SCE</em> object to a single model <em>SCE</em> object.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetCeldaList.html">subsetCeldaList</a></span><span class="op">(</span><span class="va">cgs</span>, params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fl">5</span>, L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>If the “bestOnly” parameter is set to FALSE in the <code>celdaGridSearch</code>, then the <code>selectBestModel</code> function can be used to select the chains with the lowest log likelihoods within each combination of parameters. Alternatively, users can select a specific chain by specifying the index within the <code>subsetCeldaList</code> function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celdaGridSearch.html">celdaGridSearch</a></span><span class="op">(</span><span class="va">simsce</span>,
    paramsTest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span>, L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">9</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span>,
    cores <span class="op">=</span> <span class="fl">1</span>,
    model <span class="op">=</span> <span class="st">"celda_CG"</span>,
    nchains <span class="op">=</span> <span class="fl">2</span>,
    maxIter <span class="op">=</span> <span class="fl">100</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span>,
    bestOnly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">cgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resamplePerplexity.html">resamplePerplexity</a></span><span class="op">(</span><span class="va">cgs</span>, celdaList <span class="op">=</span> <span class="va">cgs</span>, resample <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="va">cgsK5L10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetCeldaList.html">subsetCeldaList</a></span><span class="op">(</span><span class="va">cgs</span>, params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>K <span class="op">=</span> <span class="fl">5</span>, L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>

<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectBestModel.html">selectBestModel</a></span><span class="op">(</span><span class="va">cgsK5L10</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="miscellaneous-utility-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#miscellaneous-utility-functions" class="anchor"></a>Miscellaneous utility functions</h1>
<p>celda also contains several utility functions for the users’ convenience.</p>
<div id="finding-the-modules-for-feature-with-featuremodulelookup" class="section level2">
<h2 class="hasAnchor">
<a href="#finding-the-modules-for-feature-with-featuremodulelookup" class="anchor"></a>Finding the modules for feature with featureModuleLookup</h2>
<p><code>featureModuleLookup</code> can be used to look up the module a specific feature was clustered to.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/featureModuleLookup.html">featureModuleLookup</a></span><span class="op">(</span><span class="va">sce</span>, feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gene_99"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Gene_99 
##       4</code></pre>
</div>
<div id="reordering-cluster-labels-with-recodeclusterz-recodeclustery" class="section level2">
<h2 class="hasAnchor">
<a href="#reordering-cluster-labels-with-recodeclusterz-recodeclustery" class="anchor"></a>Reordering cluster labels with recodeClusterZ, recodeClusterY</h2>
<p><code>recodeClusterZ</code> and <code>recodeClusterY</code> allows the user to recode the cell and feature cluster labels, respectively.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sceZRecoded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recodeClusterZ.html">recodeClusterZ</a></span><span class="op">(</span><span class="va">sce</span>,
    from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The model prior to reordering cell labels compared to after reordering cell labels:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/celdaClusters.html">celdaClusters</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fu"><a href="../reference/celdaClusters.html">celdaClusters</a></span><span class="op">(</span><span class="va">sceZRecoded</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##    
##      1  2  3  4  5
##   1  0 44  0  0  0
##   2 42  0  0  0  0
##   3  0  0 40  0  0
##   4  0  0  0 47  0
##   5  0  0  0  0 34</code></pre>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [3] Biobase_2.50.0              GenomicRanges_1.42.0       
##  [5] GenomeInfoDb_1.26.4         IRanges_2.24.1             
##  [7] S4Vectors_0.28.1            BiocGenerics_0.36.0        
##  [9] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [11] celda_1.10.0                BiocStyle_2.18.1           
## 
## loaded via a namespace (and not attached):
##   [1] bitops_1.0-6               fs_1.5.0                  
##   [3] doParallel_1.0.16          RColorBrewer_1.1-2        
##   [5] httr_1.4.2                 rprojroot_2.0.2           
##   [7] tools_4.0.4                bslib_0.2.4               
##   [9] irlba_2.3.3                utf8_1.2.1                
##  [11] R6_2.5.0                   uwot_0.1.10               
##  [13] DBI_1.1.1                  colorspace_2.0-0          
##  [15] GetoptLong_1.0.5           withr_2.4.1               
##  [17] tidyselect_1.1.0           compiler_4.0.4            
##  [19] textshaping_0.3.5          Cairo_1.5-12.2            
##  [21] assertive.properties_0.0-4 enrichR_3.0               
##  [23] DelayedArray_0.16.2        desc_1.3.0                
##  [25] labeling_0.4.2             assertive.files_0.0-2     
##  [27] bookdown_0.21              sass_0.3.1                
##  [29] scales_1.1.1               pkgdown_1.6.1             
##  [31] systemfonts_1.0.1          stringr_1.4.0             
##  [33] digest_0.6.27              rmarkdown_2.7             
##  [35] XVector_0.30.0             assertive.numbers_0.0-2   
##  [37] pkgconfig_2.0.3            htmltools_0.5.1.1         
##  [39] highr_0.8                  fastmap_1.1.0             
##  [41] GlobalOptions_0.1.2        rlang_0.4.10              
##  [43] FNN_1.1.3                  shape_1.4.5               
##  [45] farver_2.1.0               gridGraphics_0.5-1        
##  [47] jquerylib_0.1.3            generics_0.1.0            
##  [49] combinat_0.0-8             jsonlite_1.7.2            
##  [51] dplyr_1.0.5                RCurl_1.98-1.2            
##  [53] magrittr_2.0.1             GenomeInfoDbData_1.2.4    
##  [55] multipanelfigure_2.1.2     Matrix_1.3-2              
##  [57] Rcpp_1.0.6                 munsell_0.5.0             
##  [59] fansi_0.4.2                lifecycle_1.0.0           
##  [61] stringi_1.5.3              assertive.base_0.0-9      
##  [63] yaml_2.2.1                 MCMCprecision_0.4.0       
##  [65] zlibbioc_1.36.0            Rtsne_0.15                
##  [67] plyr_1.8.6                 grid_4.0.4                
##  [69] ggrepel_0.9.1              crayon_1.4.1              
##  [71] lattice_0.20-41            circlize_0.4.12           
##  [73] magick_2.7.0               ComplexHeatmap_2.6.2      
##  [75] knitr_1.31                 pillar_1.5.1              
##  [77] rjson_0.2.20               reshape2_1.4.4            
##  [79] codetools_0.2-18           glue_1.4.2                
##  [81] evaluate_0.14              data.table_1.14.0         
##  [83] BiocManager_1.30.10        png_0.1-7                 
##  [85] vctrs_0.3.6                foreach_1.5.1             
##  [87] gtable_0.3.0               purrr_0.3.4               
##  [89] clue_0.3-58                assertthat_0.2.1          
##  [91] cachem_1.0.4               ggplot2_3.3.5             
##  [93] xfun_0.22                  assertive.types_0.0-3     
##  [95] RcppEigen_0.3.3.9.1        ragg_1.1.3                
##  [97] tibble_3.1.0               iterators_1.0.13          
##  [99] memoise_2.0.0              cluster_2.1.0             
## [101] ellipsis_0.3.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Joshua Campbell, Shiyi Yang, Zhe Wang, Sean Corbett, Yusuke Koga.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
